File: Rust/Engineering/ElectroMagnetism.rs
pub struct ElectroMagnetism;

impl ElectroMagnetism {
    pub const VACUUM_PERMITTIVITY: f64 = 8.854_187_812_8e-12;
    pub const COULOMB_CONSTANT: f64 = 8.987_551_792_3e9;

    pub fn electric_field_potential_gradient(potential_difference: f64, distance: f64) -> f64 {
        -potential_difference / distance
    }

    pub fn electric_flux_density(electric_field: f64, permittivity: f64) -> f64 {
        permittivity * electric_field
    }

    pub fn absolute_permittivity(electric_flux_density: f64, electric_field: f64) -> f64 {
        electric_flux_density / electric_field
    }

    pub fn electric_dipole_moment(charge: f64, distance: f64) -> f64 {
        charge * distance
    }

    pub fn electric_polarization(electric_dipole_moment: f64, volume: f64) -> f64 {
        electric_dipole_moment / volume
    }

    pub fn electric_displacement_field(electric_field: f64, permittivity: f64) -> f64 {
        permittivity * electric_field
    }

    pub fn electric_displacement_flux(electric_displacement_field: f64, area: f64) -> f64 {
        electric_displacement_field * area
    }

    pub fn absolute_electric_potential(electric_field: f64, distance: f64) -> f64 {
        electric_field * distance
    }

    pub fn coulombs_law_force(charge_1: f64, charge_2: f64, separation_distance: f64) -> f64 {
        Self::COULOMB_CONSTANT * charge_1 * charge_2 / separation_distance.powi(2)
    }

    pub fn capacitance(charge: f64, potential_difference: f64) -> f64 {
        charge / potential_difference
    }
}

#[cfg(test)]
mod tests {
    use super::ElectroMagnetism;

    #[test]
    fn field_and_flux_density_relationships_are_consistent() {
        let field = ElectroMagnetism::electric_field_potential_gradient(12.0, 3.0);
        let flux_density = ElectroMagnetism::electric_flux_density(2.0, 5.0);
        let permittivity = ElectroMagnetism::absolute_permittivity(10.0, 2.0);

        assert!((field + 4.0).abs() < 1e-12);
        assert!((flux_density - 10.0).abs() < 1e-12);
        assert!((permittivity - 5.0).abs() < 1e-12);
    }

    #[test]
    fn dipole_and_coulomb_helpers_produce_expected_results() {
        assert!((ElectroMagnetism::electric_dipole_moment(2.0, 0.5) - 1.0).abs() < 1e-12);
        assert!((ElectroMagnetism::electric_polarization(1.0, 0.25) - 4.0).abs() < 1e-12);
        assert!(ElectroMagnetism::coulombs_law_force(1e-6, 2e-6, 0.1) > 1.0);
    }
}

File: C#/TheOpenMathLibrary.Engineering/Electromagnetism.cs
using System;

namespace TheOpenMathLibrary.Engineering
{
    /// <summary>
    /// Provides helper formulas for introductory electromagnetism calculations.
    /// </summary>
    public class Electromagnetism
    {
        /// <summary>
        /// Calculates potential gradient magnitude from electric field and potential difference.
        /// </summary>
        /// <param name="electricField">The electric field magnitude.</param>
        /// <param name="potential">The electric potential difference.</param>
        /// <returns>The potential gradient.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="potential"/> is zero.</exception>
        public static double ElectricFieldPotentialGradient(double electricField, double potential)
        {
            EnsureNonZero(potential, nameof(potential));
            return electricField / potential;
        }

        /// <summary>
        /// Calculates electric flux density from electric field and permittivity.
        /// </summary>
        /// <param name="electricField">The electric field magnitude.</param>
        /// <param name="permittivity">The permittivity of the medium.</param>
        /// <returns>The electric flux density.</returns>
        public static double ElectricFluxDensity(double electricField, double permittivity)
        {
            return electricField * permittivity;
        }

        /// <summary>
        /// Calculates absolute permittivity from electric flux density and electric field.
        /// </summary>
        /// <param name="electricFluxDensity">The electric flux density.</param>
        /// <param name="electricField">The electric field magnitude.</param>
        /// <returns>The absolute permittivity.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="electricField"/> is zero.</exception>
        public static double AbsolutePermittivity(double electricFluxDensity, double electricField)
        {
            EnsureNonZero(electricField, nameof(electricField));
            return electricFluxDensity / electricField;
        }

        /// <summary>
        /// Calculates electric dipole moment.
        /// </summary>
        /// <param name="charge">The charge magnitude.</param>
        /// <param name="distance">The separation distance.</param>
        /// <returns>The electric dipole moment.</returns>
        public static double ElectricDipoleMoment(double charge, double distance)
        {
            return charge * distance;
        }

        /// <summary>
        /// Calculates electric polarization.
        /// </summary>
        /// <param name="electricDipoleMoment">The electric dipole moment.</param>
        /// <param name="volume">The material volume.</param>
        /// <returns>The electric polarization.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="volume"/> is zero.</exception>
        public static double ElectricPolarization(double electricDipoleMoment, double volume)
        {
            EnsureNonZero(volume, nameof(volume));
            return electricDipoleMoment / volume;
        }

        /// <summary>
        /// Calculates electric displacement field from polarization and permittivity.
        /// </summary>
        /// <param name="electricPolarization">The electric polarization.</param>
        /// <param name="permittivity">The permittivity of the medium.</param>
        /// <returns>The electric displacement field.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="permittivity"/> is zero.</exception>
        public static double ElectricDisplacementField(double electricPolarization, double permittivity)
        {
            EnsureNonZero(permittivity, nameof(permittivity));
            return electricPolarization / permittivity;
        }

        /// <summary>
        /// Calculates electric displacement flux.
        /// </summary>
        /// <param name="electricDisplacementField">The electric displacement field.</param>
        /// <param name="area">The cross-sectional area.</param>
        /// <returns>The electric displacement flux.</returns>
        public static double ElectricDisplacementFlux(double electricDisplacementField, double area)
        {
            return electricDisplacementField * area;
        }

        /// <summary>
        /// Calculates absolute electric potential from field and distance.
        /// </summary>
        /// <param name="electricField">The electric field magnitude.</param>
        /// <param name="distance">The displacement distance.</param>
        /// <returns>The electric potential difference.</returns>
        public static double AbsoluteElectricPotential(double electricField, double distance)
        {
            return electricField * distance;
        }

        private static void EnsureNonZero(double value, string parameterName)
        {
            if (value == 0d)
            {
                throw new ArgumentOutOfRangeException(parameterName, "The value must not be zero.");
            }
        }
    }

    /// <summary>
    /// Backward-compatible type alias for <see cref="Electromagnetism"/>.
    /// </summary>
    public class ElectroMagnetism : Electromagnetism
    {
    }
}
